Documentation/Evolution/CharacterClasses.md

```diff
@@ -0,0 +1,481 @@
+
+We used `\(rationale)` for `Character`. (If not obvious, reasons).
+
+</details>
+
+
+
+### Unicode named values and properties: `\N`, `\p`, `\P`
+
+`\N{NAME}` matches a Unicode scalar value with the specified name. `\p{PROPERTY}` and `\p{PROPERTY=VALUE}` match a Unicode scalar value with the given Unicode property (and value, if given). 
+
+While most Unicode-defined properties can only match at the Unicode scalar level, some are defined to match an extended grapheme cluster. For example, `/\p{RGI_Emoji_Flag_Sequence}/` will match any flag emoji character, which are composed of two Unicode scalar values.
+
+`\P{...}` matches the inverse of `\p{...}`.
+
+Most of this functionality is already provided inside `Unicode.Scalar.Properties`, and we propose to round out Swift's current support with:
+
+```swift
+// TODO: any that are missing
+```
+
+**TODO**: Check with Alejandro that the code size impact is reasonable
+
+Even though we are not proposing any `Character`-based API, we'd like to discuss with the community whether or how to extend them to grapheme clusters. Some options:
+
+- Forbid in any grapheme-cluster semantic mode
+- Match only single-scalar grapheme clusters with the given property
+- Match any grapheme cluster that starts with the given property
+- Something more-involved such as per-property reasoning
+
+
+### POSIX character classes: `[:NAME:]`
+
+We propose that POSIX character classes be named "posixName" with the following semantics:
+
+```swift
+extension Unicode.Scalar {
+  /// TODO:
+  public var isPOSIXAlphanumeric: Bool { get }
+
+  /// TODO:
+  public var isPOSIXAlphabetic: Bool { get }
+
+  /// TODO:
+  public var isASCII: Bool { get } // if not 
+
+  /// TODO:
+  public var isPOSIXBlank: Bool { get }
+
+  /// TODO:
+  public var isPOSIXControl: Bool { get }
+
+  /// TODO:
+  public var isPOSIXDigit: Bool { get }
+
+  /// TODO:
+  public var isPOSIXGraph: Bool { get }
+
+  /// TODO:
+  public var isPOSIXLowercase: Bool { get }
+
+  /// TODO:
+  public var isPOSIXPrint: Bool { get }
+
+  /// TODO:
+  public var isPOSIXPunctuation: Bool { get }
+
+  /// TODO:
+  public var isPOSIXSpace: Bool { get }
+
+  /// TODO:
+  public var isPOSIXUppercase: Bool { get }
+
+  /// TODO:
+  public var isPOSIXWord: Bool { get }
+
+  /// TODO:
+  public var isPOSIXHexDigit: Bool { get }
+}
+// ... same for Character ...
+```
+
+`Unicode.Scalar.isASCII` already exists and `Character.isASCII` and can satisfy `[:ascii:]`.
+
+Alternatively, we could introduce an option-set-like `POSIXCharacterClass` and `func isPOSIX(_:POSIXCharacterClass)` since POSIX is a fully defined standard. This would cut down on the amount of API noise directly visible on `Character` and `Unicode.Scalar` significantly.
+
+We'd like some more discussion with the community here, and it's possible this will become clearer as more of the string processing story takes shape.
+
```

Comment:
Maybe we just strike the `public` additions to `Unicode.Scalar`. Let's give a treatment to which of these are not trivial compositions of `isFoo && isASCII` and which are meaningfully different. Then we can talk about adding `isPOSIXFoo` vs the option set approach. The option set approach might be `POSIXCharacterClass` or `ASCIICharacterClass`, ideally the latter if we want to bless ASCII-mode matching (which we mostly likely will want to do anyways).